### PR TITLE
feat(logging): update the current logging implementation

### PR DESCRIPTION
**Pull request by hamzamahmood**
## What
This PR updates the current implementation of the logging and logging configuration and some code cleanup changes.

- We renamed HttpLogger to SdkLogger
- We added ConsoleLogger for logging to console without logging providers
- We added NullLogger that implements ApiLogger
- We updated SdkLogger to only have logRequest and logResponse methods only
- We updated ApiLoggingConfiguration to now have multi-level configuration for request and response
- We added getUrl() method in Request interface that returns URL without query params
- We updated core-interfaces-java and okhttp-client-adapter package versions

## Why
The following changes were made to make the Java logging use structured logging and improve user experience.

## Type of change
Select multiple if applicable.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause a breaking change)
- [ ] Tests (adds or updates tests)
- [ ] Documentation (adds or updates documentation)
- [ ] Refactor (style improvements, performance improvements, code refactoring)
- [ ] Revert (reverts a commit)
- [ ] CI/Build (adds or updates a script, change in external dependencies)

## Dependency Change
Updated core-interfaces-java and okhttp-client-adapter package versions

## Breaking change
Updated ApiLoggingConfiguration

## Checklist
- [x] My code follows the coding conventions
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added new unit tests

### CODE REVIEW

**Review by sufyankhanrao — CHANGES_REQUESTED** [review]

File: src/test/java/apimatic/core/EndToEndTest.java [hunk 1]
```diff
@@ -500,7 +501,9 @@ private GlobalConfiguration getGlobalConfig() {
                 .compatibilityFactory(getCompatibilityFactory()).httpClient(httpClient)
                 .baseUri(server -> getBaseUri(server)).callback(callback).userAgent(userAgent)
                 .userAgentConfig(Collections.emptyMap()).additionalHeaders(null)
-                .globalHeader("version", "0.1").globalHeader("version", "1.2").build();
+                .globalHeader("version", "0.1").globalHeader("version", "1.2")
+                .loggingConfiguration(new ApiLoggingConfiguration.Builder().build())
```
[the remark this answers is not in the document]
Comment (hamzamahmood) [reply src/test/java/apimatic/core/EndToEndTest.java]: loggingConfiguration has been removed from here.

File: src/test/java/apimatic/core/configurations/http/client/ApiLoggingConfigurationTest.java [hunk 2]
```diff
@@ -1,81 +1,20 @@
...
-import static org.junit.Assert.assertTrue;
 import java.util.TreeSet;
 import org.junit.Test;
+import org.slf4j.event.Level;
+
 import io.apimatic.core.configurations.http.client.ApiLoggingConfiguration;
-import io.apimatic.coreinterfaces.http.LoggingLevel;
-import io.apimatic.coreinterfaces.http.LoggingPolicy;
 
 public class ApiLoggingConfigurationTest {
```
[the remark this answers is not in the document]
Comment (hamzamahmood) [reply src/test/java/apimatic/core/configurations/http/client/ApiLoggingConfigurationTest.java]: This file has been removed

File: src/main/java/io/apimatic/core/logger/ConsoleLogger.java [hunk 3]
```diff
@@ -18,9 +18,8 @@ public class ConsoleLogger implements Logger {
     @Override
     public void log(Level level, String format, Map<String, Object> arguments) {
         StringBuilder builder = new StringBuilder();
-        builder.append(level);
-        builder.append(": ");
-        builder.append(MessageFormatter.basicArrayFormat(format, arguments.values().toArray()));
+        builder.append(String.format("%s: %s", level,
```
Comment (sufyankhanrao) [inline src/main/java/io/apimatic/core/logger/ConsoleLogger.java:21]: You can get rid of this StringBuilder as well.

Comment (hamzamahmood) [reply src/main/java/io/apimatic/core/logger/ConsoleLogger.java]: we are now directly using String Format in system out print statement

File: src/test/java/apimatic/core/logger/LoggerUtilitiesTest.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        headers.put("Masked-Header", "MaskedValue");
+
+        List<String> headersToWhiteList = getLowerCaseList("Accept");
+
+        Map<String, String> extractedHeaders = LoggerUtilities.filterSensitiveHeaders(headers,
+                headersToWhiteList, enableMasking);
+
+        final int expectedHeaderSize = 2;
+        final String expectedAcceptHeaderValue = "Value1";
+        final String expectedPoweredByHeaderValue = "**Redacted**";
```
Comment (sufyankhanrao) [inline src/test/java/apimatic/core/logger/LoggerUtilitiesTest.java:84]: `expectedPoweredByHeaderValue` should be renamed to `expectedMaskedHeaderValue`

File: src/test/java/apimatic/core/logger/LoggerUtilitiesTest.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        final int expectedHeaderSize = 3;
+        assertEquals(expectedHeaderSize, extractedHeaders.size());
+        assertTrue(extractedHeaders.containsKey("Header1"));
+        assertTrue(extractedHeaders.containsKey("Header2"));
+        assertTrue(extractedHeaders.containsKey("Header3"));
+    }
+
+    @Test
+    public void testExtractHeadersToLogWithFilter() {
+        final boolean enableMasking = true;
```
Comment (sufyankhanrao) [inline src/test/java/apimatic/core/logger/LoggerUtilitiesTest.java:72]: There should be another test case that use the enableMasking value as `false`.

File: src/main/java/io/apimatic/core/logger/LoggerUtilities.java [hunk 6]
```diff
@@ -43,19 +50,20 @@ public static Map<String, String> extractHeadersToLog(Map<String, String> header
...
             Map<String, String> filteredHeaders = new HashMap<>();
             for (Map.Entry<String, String> entry : headers.entrySet()) {
                 String key = entry.getKey().toLowerCase();
-                if (SENSITIVE_HEADERS.contains(key)) {
-                    filteredHeaders.put(entry.getKey(), "**Redacted**");
-                } else {
+                if (NON_SENSITIVE_HEADERS.contains(key) || headersToWhiteList.contains(key)) {
                     filteredHeaders.put(entry.getKey(), entry.getValue());
+                } else {
+                    filteredHeaders.put(entry.getKey(), "**Redacted**");
```
[the remark this answers is not in the document]
Comment (hamzamahmood) [reply src/main/java/io/apimatic/core/logger/LoggerUtilities.java:67]: Yes, this is the correct behavior. I.e. all headers not in NON_SENSITIVE_HEADERS list should be marked as Redacted until user unmask it using the configuration.

File: src/test/java/apimatic/core/logger/LoggerUtilitiesTest.java (hunk 5 above)
Comment (hamzamahmood) [reply src/test/java/apimatic/core/logger/LoggerUtilitiesTest.java:72]: test case added

File: src/test/java/apimatic/core/logger/LoggerUtilitiesTest.java (hunk 4 above)
Comment (hamzamahmood) [reply src/test/java/apimatic/core/logger/LoggerUtilitiesTest.java]: variable has been renamed

File: src/main/java/io/apimatic/core/logger/configurations/LogBaseOptions.java [hunk 7]
```diff
@@ -13,6 +13,7 @@ public abstract class LogBaseOptions {
     private boolean logHeaders = false;
     private List<String> headersToInclude = new ArrayList<>();
     private List<String> headersToExclude = new ArrayList<>();
+    private List<String> headersToWhiteList = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (hamzamahmood) [reply src/main/java/io/apimatic/core/logger/configurations/LogBaseOptions.java]: unmaskHeaders added to Request and Response Log options

File: src/main/java/io/apimatic/core/logger/LoggerUtilities.java [hunk 8]
```diff
@@ -26,6 +28,28 @@ public final class LoggerUtilities {
...
+     * headers, and sensitivity masking configuration.
+     * @param loggingOptions       The logging options containing configurations for
+     *                             header logging.
+     * @param headers              The headers to be evaluated for logging.
+     * @param maskSensitiveHeaders Determines whether sensitive headers should be
+     *                             masked in the log.
+     * @return A map containing the headers to be logged, considering the provided
+     *         options and sensitivity masking.
+     */
+    public static Map<String, String> getHeadersToLog(ReadonlyLogOptions loggingOptions,
```
Comment (sufyankhanrao) [inline src/main/java/io/apimatic/core/logger/LoggerUtilities.java:42]: Do we need test cases for this method? If yes, kindly consider adding them.

Comment (sufyankhanrao) [reply src/main/java/io/apimatic/core/logger/LoggerUtilities.java:42]: Ok as long as it is being covered then looks okay.

File: src/main/java/io/apimatic/core/logger/SdkLogger.java [hunk 9]
```diff
@@ -58,22 +58,20 @@ public SdkLogger(final LoggingConfiguration config) {
...
         String queryParameter = CoreHelper.getQueryParametersFromUrl(request.getQueryUrl());
         String contentType = request.getHeaders().value(LoggerConstants.CONTENT_TYPE) != null
                 ? request.getHeaders().value(LoggerConstants.CONTENT_TYPE)
                 : "";
 
         Map<String, Object> requestArguments = new LinkedHashMap<String, Object>();
         requestArguments.put(LoggerConstants.METHOD, request.getHttpMethod());
-        requestArguments.put(LoggerConstants.URL, url);
-        requestArguments.put(LoggerConstants.CONTENT_TYPE, contentType);
+        requestArguments.put(LoggerConstants.URL, request.getUrl());
```
Comment (sufyankhanrao) [inline src/main/java/io/apimatic/core/logger/SdkLogger.java:68]: Nits, Non-blocker: We can minimize the number of lines by putting the decision on queryParams in Path.
```Java
requestArguments.put(LoggerConstants.URL, getReqestUrl(request));

// following is the method implementation
private string getRequestUrl(Request request) {
    if (requestLogOptions.shouldIncludeQueryInPath()) {
        return request.getQueryUrl();
    }

    return request.getUrl();
}
```

Comment (hamzamahmood) [reply src/main/java/io/apimatic/core/logger/SdkLogger.java]: Resolved by extracting the logic inside a private method

File: README.md [hunk 10]
```diff
@@ -13,42 +13,50 @@ This project contains core logic and the utilities for the APIMatic's Java SDK
...
+| [`Slf4jLogger`](./src/main/java/io/apimatic/core/logger/Slf4jLogger.java)                                                        | Logger implementation supporting Slf4j Facade.                                                                                                                                                                                                        |
+| [`ConsoleLogger`](./src/main/java/io/apimatic/core/logger/ConsoleLogger.java)                                                    | Default Logger implementation if logging is enabled.                                                                                                                                                                                                  |
+| [`LoggerFactory`](./src/main/java/io/apimatic/core/logger/LoggerFactory.java)                                                    | Returns an instance of ApiLogger based on the provided configuration.                                                                                                                                                                                 |
```
Comment (sufyankhanrao) [inline README.md:28]: The link for LoggerFactory is broken and should be fixed.
image

File: src/main/java/io/apimatic/core/logger/LoggerUtilities.java [hunk 11]
```diff
@@ -31,37 +32,37 @@ private LoggerUtilities() {
...
      * Extracts headers to log based on inclusion and exclusion criteria.
-     *
      * @param headers          The map of headers.
      * @param headersToInclude The set of headers to include.
      * @param headersToExclude The set of headers to exclude.
      * @return The extracted headers to log.
      */
     public static Map<String, String> extractHeadersToLog(Map<String, String> headers,
-            List<String> headersToInclude, List<String> headersToExclude) {
+                                                          List<String> headersToInclude,
```
Comment (sufyankhanrao) [inline src/main/java/io/apimatic/core/logger/LoggerUtilities.java:63]: I don't see this convention in Java normally but is this a recommendation from the linter or you just read about it somewhere?

File: README.md (hunk 10 above)
Comment (hamzamahmood) [reply README.md:28]: Broken link has been resolved

File: src/main/java/io/apimatic/core/logger/LoggerUtilities.java (hunk 11 above)
Comment (hamzamahmood) [reply src/main/java/io/apimatic/core/logger/LoggerUtilities.java:63]: This might be due to my local code formatter settings, I have reverted the changes.

**Review by sufyankhanrao — DISMISSED** [review]

**Review by sufyankhanrao — APPROVED** [review]

**Pull request merged**